core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -366,66 +371,90 @@ protected boolean canLoad(Object key) {
-                     } else {
-                        cf.complete(ice);
-                     }
-                  });
-               } else {
-                  if (trace) {
-                     log.tracef("Missed entry load for key %s from store", key);
-                  }
-                  if (getStatisticsEnabled()) {
-                     cacheMisses.incrementAndGet();
-                  }
-                  cf.complete(null);
-               }
-            } catch (Throwable outerT) {
-               cf.completeExceptionally(outerT);
-            }
-         });
-      }
-
-      // If another thread is completing the request, then resume on a different CPU thread so we don't have to
-      // wait until the other command completes
-      return otherCF.thenAcceptAsync(entry -> {
+      Consumer<? super InternalCacheEntry<K, V>> action = entry -> {
          if (entry != null) {
             entryFactory.wrapExternalEntry(ctx, key, entry, true, cmd instanceof WriteCommand);
          }
          CacheEntry contextEntry = ctx.lookupEntry(key);
          if (contextEntry instanceof MVCCEntry) {
             ((MVCCEntry) contextEntry).setLoaded(true);
          }
-      }, cpuExecutor);
+      };
+
+      // If another thread is completing the request, then resume on a different CPU thread so we don't have to
+      // wait until the other command completes
+      if (otherCF != null) {
+         if (trace) {
+            log.tracef("Piggybacking on concurrent cache loader for key %s", key);
+         }
+         return otherCF.thenAcceptAsync(action, cpuExecutor);
+      }
+
+      int segment = SegmentSpecificCommand.extractSegment(cmd, key, partitioner);
+      CompletionStage<InternalCacheEntry<K, V>> result = loadAndStoreInDataContainer(ctx, key, segment, cmd);
+      result.whenComplete((value, throwable) -> {
+         // Make sure we clean up our pendingLoads properly and before completing any responses
+         pendingLoads.remove(key);
+         if (throwable != null) {
+            cf.completeExceptionally(throwable);
+         } else {
+            cf.complete(value);
+         }
+      });
+      return cf.thenAccept(action);
+   }
+
+   public CompletionStage<InternalCacheEntry<K, V>> loadAndStoreInDataContainer(InvocationContext ctx, Object key,
+                                                                                int segment, FlagAffectedCommand cmd) {
+      InternalCacheEntry<K, V> entry = dataContainer.peek(segment, key);
+      boolean includeStores = true;
+      if (entry != null) {
+         if (!entry.canExpire() || !entry.isExpired(timeService.wallClockTime())) {
+            return CompletableFuture.completedFuture(entry);
+         }
+         includeStores = false;
+      }
+
+      if (trace) {
+         log.tracef("Loading entry for key %s", key);
+      }
+      CompletionStage<InternalCacheEntry<K, V>> resultStage = PersistenceUtil.<K, V>loadEntryAsync(persistenceManager, key,
```

Comment:
`loadEntryAsync` is only ever used in this class, should we make it a local method?

Reply:
Same with `putIfAbsentDataContainer`

Reply:
Originally this wasn't due to `SEGMENT_NOT_PROVIDED` but I believe I have changed this now so that isn't a possible value any longer. So I can move them over.